src/Npgsql.GeoJSON/GeoJsonHandler.cs

```diff
@@ -0,0 +1,796 @@
+        LongCrs
+    }
+
+    public class GeoJsonHandlerFactory : NpgsqlTypeHandlerFactory<GeoJSONObject>
+    {
+        internal readonly GeoJsonOption _option;
+
+        public GeoJsonHandlerFactory(GeoJsonOption option = GeoJsonOption.None)
+            => _option = option;
+
+        protected override NpgsqlTypeHandler<GeoJSONObject> Create(NpgsqlConnection conn)
+        {
+            var builder = new CrsMapBuilder();
+            if ((_option & (GeoJsonOption.ShortCrs | GeoJsonOption.LongCrs)) != GeoJsonOption.None)
+                using (var cmd = new NpgsqlCommand(
+                    "SELECT min(srid), max(srid), auth_name " +
+                    "FROM(SELECT srid, auth_name, srid - rank() OVER(ORDER BY srid) AS range " +
+                    "FROM spatial_ref_sys) AS s GROUP BY range, auth_name ORDER BY 1;", conn))
+                using (var reader = cmd.ExecuteReader())
+                    while (reader.Read())
+                    {
+                        builder.Add(new CrsMapEntry(
+                            reader.GetInt32(0),
+                            reader.GetInt32(1),
+                            reader.GetString(2)));
+                    }
+            return new GeoJsonHandler(_option, builder.Build());
+        }
+    }
+
+    sealed class GeoJsonHandler : NpgsqlTypeHandler<GeoJSONObject>,
+        INpgsqlTypeHandler<Point>, INpgsqlTypeHandler<MultiPoint>,
+        INpgsqlTypeHandler<Polygon>, INpgsqlTypeHandler<MultiPolygon>,
+        INpgsqlTypeHandler<LineString>, INpgsqlTypeHandler<MultiLineString>,
+        INpgsqlTypeHandler<GeometryCollection>,
+        INpgsqlTypeHandler<IGeoJSONObject>,
+        INpgsqlTypeHandler<IGeometryObject>
+    {
+        readonly GeoJsonOption _option;
+        readonly CrsMap _crsMap;
+
+        internal GeoJsonHandler(GeoJsonOption option, CrsMap crsMap)
+        {
+            _option = option;
+            _crsMap = crsMap;
+        }
+
+        GeoJsonOption CrsType => _option & (GeoJsonOption.ShortCrs | GeoJsonOption.LongCrs);
+
+        bool BoundingBox => (_option & GeoJsonOption.BoundingBox) != 0;
+
+        static bool HasSrid(EwkbGeometryType type)
+            => (type & EwkbGeometryType.HasSrid) != 0;
+
+        static bool HasZ(EwkbGeometryType type)
+            => (type & EwkbGeometryType.HasZ) != 0;
+
+        static bool HasM(EwkbGeometryType type)
+            => (type & EwkbGeometryType.HasM) != 0;
+
+        static bool HasZ(IPosition coordinates)
+            => coordinates.Altitude.HasValue;
+
+        const int SizeOfLength = sizeof(int);
+        const int SizeOfHeader = sizeof(byte) + sizeof(EwkbGeometryType);
+        const int SizeOfHeaderWithLength = SizeOfHeader + SizeOfLength;
+        const int SizeOfPoint2D = 2 * sizeof(double);
+        const int SizeOfPoint3D = 3 * sizeof(double);
+
+        static int SizeOfPoint(bool hasZ)
+            => hasZ ? SizeOfPoint3D : SizeOfPoint2D;
+
+        static int SizeOfPoint(EwkbGeometryType type)
```

Comment:
Maybe just do a switch? Would be faster (even if it's surely negligible)

Reply:
Yep.

Reply:
I did some perf measurements out of curiosity and got the following result:

```csharp
[Config(typeof(Config))]
public class SwithcVsIf
{
    private EwkbGeometryType _type;

    [Params(
        (uint)(EwkbGeometryType.Point),
        (uint)(EwkbGeometryType.Point | EwkbGeometryType.HasZ),
        (uint)(EwkbGeometryType.Point | EwkbGeometryType.HasM),
        (uint)(EwkbGeometryType.Point | EwkbGeometryType.HasZ | EwkbGeometryType.HasM))]
    public uint Type
    {
        get => (uint)_type;
        set => _type = (EwkbGeometryType)value;
    }

    public const int Iterations = 1000000;

    [Benchmark(OperationsPerInvoke = Iterations)]
    public int UsingSwitch()
    {
        var total = 0;
        for (var i = 0; i < Iterations; i++)
            total += SizeOf(_type);
        return total;

        int SizeOf(EwkbGeometryType type)
        {
            switch (type & (EwkbGeometryType.HasZ | EwkbGeometryType.HasM))
            {
                case EwkbGeometryType.HasZ:
                case EwkbGeometryType.HasM:
                    return sizeof(double) * 3;

                case EwkbGeometryType.HasZ | EwkbGeometryType.HasM:
                    return sizeof(double) * 4;

                default:
                    return sizeof(double) * 2;
            }
        }
    }

    [Benchmark(OperationsPerInvoke = Iterations)]
    public int UsingIf()
    {
        var total = 0;
        for (var i = 0; i < Iterations; i++)
            total += SizeOf(_type);
        return total;

        int SizeOf(EwkbGeometryType type)
        {
            var size = sizeof(double) * 2;
            if (HasZ(type))
                size += sizeof(double);
            if (HasM(type))
                size += sizeof(double);
            return size;
        }
    }
```

```
BenchmarkDotNet=v0.10.12, OS=Windows 10.0.17133
Intel Xeon CPU E5-2687W 0 3.10GHz, 1 CPU, 16 logical cores and 16 physical cores
.NET Core SDK=2.1.4
  [Host]     : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT
```

|      Method |       Type |     Mean |     Error |    StdDev |   Median |          Op/s |
|------------ |----------- |---------:|----------:|----------:|---------:|--------------:|
| UsingSwitch |          1 | 7.840 ns | 0.1837 ns | 0.4873 ns | 7.765 ns | 127,548,559.1 |
|     UsingIf |          1 | 9.434 ns | 0.1880 ns | 0.5017 ns | 9.484 ns | 105,994,853.5 |
| UsingSwitch | 1073741825 | 7.214 ns | 0.4103 ns | 1.2097 ns | 7.703 ns | 138,619,873.1 |
|     UsingIf | 1073741825 | 6.440 ns | 0.5014 ns | 1.4784 ns | 7.073 ns | 155,286,533.5 |
| UsingSwitch | 2147483649 | 7.086 ns | 0.4545 ns | 1.3402 ns | 7.521 ns | 141,114,206.3 |
|     UsingIf | 2147483649 | 7.781 ns | 0.4938 ns | 1.4561 ns | 8.360 ns | 128,510,638.6 |
| UsingSwitch | 3221225473 | 8.803 ns | 0.1749 ns | 0.4576 ns | 8.775 ns | 113,593,669.6 |
|     UsingIf | 3221225473 | 6.921 ns | 0.1329 ns | 0.2528 ns | 6.928 ns | 144,487,319.6 |

I think it's more readability choice because the difference is very small. It makes sense to leave it as it is.